Modules/_ctypes/_ctypes.c

```diff
@@ -710,13 +711,18 @@ StructUnionType_init(PyObject *self, PyObject *args, PyObject *kwds, int isStruc
         if (baseinfo == NULL) {
             return 0;
         }
-
+        int ret = 0;
+        STGINFO_LOCK(baseinfo);
         /* copy base info */
-        if (PyCStgInfo_clone(info, baseinfo) < 0) {
-            return -1;
+        ret = PyCStgInfo_clone(info, baseinfo);
+        if (ret >= 0) {
+            // clear the 'final' bit in the subclass info
+            // safe to modify without atomics as it is not exposed to other threads
+            info->dict_final = 0;
```

Comment:
I think this would be less confusing if the clearing of `dict_final` happens in `PyCStgInfo_clone` (just like we reset the mutex).

This is the only caller of `PyCStgInfo_clone` so I think it's functionally equivalent.